File: oxidize-python/oxidize_python/internal/generate/stream.py
"""Token streaming for server SSE mirroring oxidize-golang/internal/generate/stream.go."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

from oxidize_python.core.model.generation import GenerationStream, default_generation_config
from oxidize_python.core.model.inference import InferenceModel
from oxidize_python.core.model.loader import LoaderConfig
from oxidize_python.core.model.model import Session, Token
from oxidize_python.core.tokenizer import from_gguf_metadata
from oxidize_python.core.tokenizer.bpe import BpeTokenizer
from oxidize_python.core.tokenizer.tokenizer import EncodeOptions, SpecialTokens
from oxidize_python.internal.generate import PlaceholderSpec, placeholder_text
from oxidize_python.internal.generate.cache import inference_from_cache
from oxidize_python.internal.gguf.parse import load_file


@dataclass
class CompletionParams:
    max_tokens: int = 64
    temperature: float = 0.8
    top_p: float = 0.9
    top_k: int = 0
    loader: LoaderConfig | None = None


TokenHandler = Callable[[str], None]


def _load_tokenizer(model_path: str):
    try:
        gguf = load_file(model_path)
        meta = {k: v.string for k, v in gguf.metadata.items() if v.string}
        return from_gguf_metadata(meta)
    except Exception:
        return BpeTokenizer([], [], SpecialTokens(bos=1, eos=2))


def stream_completion(
    model_path: str,
    prompt: str,
    params: CompletionParams,
    on_token: TokenHandler | None = None,
) -> str:
    prompt = prompt.strip()
    if not prompt:
        return ""
    if params.max_tokens <= 0:
        params.max_tokens = 64
    loader = params.loader or LoaderConfig()

    try:
        inference, _entry = inference_from_cache(model_path, loader)
    except Exception:
        text = placeholder_text(PlaceholderSpec()) or ""
        if on_token and text:
            on_token(text)
        return text

    if (
        not isinstance(inference, InferenceModel)
        or inference.stack is None
        or not inference.stack.loaded()
    ):
        text = placeholder_text(PlaceholderSpec()) or ""
        if on_token and text:
            on_token(text)
        return text

    tok = _load_tokenizer(model_path)
    prompt_tokens = tok.encode(prompt, EncodeOptions()) or [Token(1)]

    session = Session()
    gen_cfg = default_generation_config()
    gen_cfg.max_new_tokens = params.max_tokens
    gen_cfg.sampling.temperature = params.temperature
    gen_cfg.sampling.top_p = params.top_p
    if params.top_k > 0:
        gen_cfg.sampling.top_k = params.top_k
    if params.temperature <= 0:
        gen_cfg.sampling.temperature = 0.0

    stream = GenerationStream(inference, session, gen_cfg)
    stream.seed(prompt_tokens)

    parts: list[str] = []
    for _ in range(params.max_tokens):
        token, done, err = stream.next()
        if err is not None:
            break
        if done:
            break
        try:
            piece = tok.decode([token])
        except Exception:
            piece = f"<{token}>"
        parts.append(piece)
        if on_token:
            on_token(piece)
    return "".join(parts)

File: oxidize-golang/internal/generate/stream.go
package generate

import (
	"context"
	"fmt"
	"strings"

	"github.com/Zapdev-labs/oxidize/golang/core/model"
	"github.com/Zapdev-labs/oxidize/golang/core/tokenizer"
)

// TokenHandler receives decoded text pieces during streaming generation.
type TokenHandler func(piece string) error

// StreamCompletion generates text token-by-token using the model cache.
func StreamCompletion(ctx context.Context, modelPath, prompt string, params CompletionParams, onToken TokenHandler) (string, error) {
	prompt = strings.TrimSpace(prompt)
	if prompt == "" {
		return "", nil
	}
	if params.MaxTokens <= 0 {
		params.MaxTokens = 64
	}

	inference, entry, err := InferenceFromCache(modelPath, params.Loader)
	if err != nil {
		return "", err
	}
	if inference.Stack == nil || !inference.Stack.Loaded() {
		text := PlaceholderText(PlaceholderSpec{})
		if onToken != nil && text != "" {
			_ = onToken(text)
		}
		return text, nil
	}
	_ = entry

	tok, err := tokenizer.LoadFromGGUFFile(modelPath)
	if err != nil {
		tok = tokenizer.NewBpeTokenizer(nil, nil, tokenizer.SpecialTokens{BOS: 1, EOS: 2})
	}
	promptTokens, err := tok.Encode(prompt, tokenizer.EncodeOptions{})
	if err != nil {
		return "", fmt.Errorf("encode: %w", err)
	}
	if len(promptTokens) == 0 {
		promptTokens = []model.Token{1}
	}

	session := model.NewSession()
	genCfg := model.DefaultGenerationConfig()
	genCfg.MaxNewTokens = params.MaxTokens
	genCfg.Sampling.Temperature = params.Temperature
	genCfg.Sampling.TopP = params.TopP
	if params.TopK > 0 {
		genCfg.Sampling.TopK = params.TopK
	}
	if params.Temperature <= 0 {
		genCfg.Sampling.Temperature = 0
	}

	stream := model.NewGenerationStream(inference, session, genCfg)
	stream.Seed(promptTokens)

	var out strings.Builder
	for i := 0; i < params.MaxTokens; i++ {
		if err := ctx.Err(); err != nil {
			return out.String(), err
		}
		token, done, err := stream.Next(ctx)
		if err != nil {
			return out.String(), err
		}
		if done {
			break
		}
		piece, err := tok.Decode([]model.Token{token})
		if err != nil {
			piece = fmt.Sprintf("<%d>", token)
		}
		out.WriteString(piece)
		if onToken != nil {
			if err := onToken(piece); err != nil {
				return out.String(), err
			}
		}
	}
	return out.String(), nil
}
